contracts/oracle/src/lib.rs

```diff
@@ -149,18 +149,25 @@ impl Contract {
             reference_hash: None,
         };
 
-        self.used_identities.insert(&external_id);
-        ext_registry::ext(self.registry.clone())
-            .with_attached_deposit(MINT_COST_REG)
-            .with_static_gas(Gas::ONE_TERA * 6)
-            .sbt_mint(vec![(claim.claimer, vec![metadata])]);
-
-        // TODO: add callback to undo identity insert if the minting didn't work because of a duplicate mint.
         if let Some(memo) = memo {
             env::log_str(&format!("SBT mint memo: {}", memo));
         }
 
-        Ok(())
+        let result = ext_registry::ext(self.registry.clone())
+            .with_attached_deposit(MINT_COST_REG)
+            .with_static_gas(Gas::ONE_TERA * 6)
+            .sbt_mint(vec![(claim.claimer, vec![metadata])]).then(Self::ext(env::current_account_id()).sbt_mint_callback(&external_id));
+
+        Ok(result)
+    }
+
+    #[private]
+    #[handle_result]
+    pub fn sbt_mint_callback(&mut self, external_id: &Vec<u8>, #[callback_result] last_result: Result<TokenId, near_sdk::PromiseError>) -> Result<TokenId, near_sdk::PromiseError> {
+        if last_result.is_ok() {
+            self.used_identities.insert(&external_id);
+        }
```

Comment:
the old form was correct this one is not. 
Once the `insert` is done in the `sbt_mint`, we know that external_id can't be used in the meantime, so no reentrency is possible.
However now, the reentrency is possible.

Reply:
Ok, I misunderstood the docs then. Rolled back the changes